File: typescript/src/core/xbrl/queries.ts
/**
 * Fact-level query primitives over the company-facts payload.
 *
 * Pure functions that normalize raw company-facts entries into typed
 * fact records; no I/O happens in this module.
 */

import { XbrlFact } from "../../types/xbrl";

/** Query result: an array of fact records with chainable helpers. */
export class FactQuery extends Array<XbrlFact> {
  static get [Symbol.species](): ArrayConstructor {
    return Array;
  }

  static fromRecords(records: Iterable<XbrlFact>): FactQuery {
    const query = new FactQuery();
    for (const record of records) {
      query.push(record);
    }
    return query;
  }

  /** Narrow the results to a single concept name (substring match). */
  byConcept(concept: string): FactQuery {
    const needle = concept.toLowerCase();
    return FactQuery.fromRecords(
      this.filter((record) =>
        String(record.concept || "")
          .toLowerCase()
          .includes(needle),
      ),
    );
  }
}

export interface ParsedFilterExpression {
  concept?: string;
  taxonomy?: string;
  unit?: string;
  period?: string;
}

/** Parse a "concept=Assets&unit=USD" filter expression. */
export function parseFilterExpression(
  expression: string,
): ParsedFilterExpression {
  const parsed: Record<string, string> = {};
  for (const part of expression.split("&")) {
    const eq = part.indexOf("=");
    if (eq > 0) {
      parsed[part.slice(0, eq).trim()] = part.slice(eq + 1).trim();
    }
  }
  return parsed;
}

/** Normalize one concept's raw company-facts entries into fact records. */
export function buildFactRecords(
  conceptName: string,
  conceptData: Record<string, any>,
  taxonomy: string,
  unitFilter?: string,
  periodFilter?: string,
): XbrlFact[] {
  const results: XbrlFact[] = [];

  const units = conceptData.units || {};
  for (const [unit, unitData] of Object.entries(units)) {
    if (unitFilter && unit !== unitFilter) {
      continue;
    }
    if (!Array.isArray(unitData)) {
      continue;
    }

    for (const fact of unitData) {
      if (periodFilter) {
        const matchesEndDate = fact.end === periodFilter;
        const matchesInstant = fact.instant === periodFilter;
        const factPeriod = fact.fy || fact.fp || fact.frame || "";
        const matchesPeriod = String(factPeriod).includes(periodFilter);
        if (!matchesEndDate && !matchesInstant && !matchesPeriod) {
          continue;
        }
      }

      results.push({
        concept: conceptName,
        taxonomy,
        value: fact.val,
        unit,
        period: fact.frame || `FY${fact.fy || ""}${fact.fp || ""}`,
        fiscal_year: fact.fy,
        fiscal_period: fact.fp,
        start_date: fact.start,
        end_date: fact.end,
        period_end: fact.start ? fact.end : undefined,
        period_instant: fact.start ? undefined : fact.end,
        context: fact.accn,
        filed: fact.filed,
        accession_number: fact.accn,
        form: fact.form,
      });
    }
  }

  return results;
}

export interface FactHistoryRow {
  concept: string;
  value: number | string | null;
  unit: string;
  periodEnd: string | null;
  periodInstant: string | null;
  end: string | null;
  filed: string | null;
  form: string | null;
}

/** History of one concept from a raw company-facts payload, sorted by end. */
export function factsHistory(
  facts: Record<string, any>,
  concept: string,
  taxonomy: string = "us-gaap",
): FactHistoryRow[] {
  const conceptData = facts?.facts?.[taxonomy]?.[concept];
  if (!conceptData) {
    return [];
  }

  const rows: FactHistoryRow[] = [];
  for (const [unit, unitFacts] of Object.entries(conceptData.units || {})) {
    if (!Array.isArray(unitFacts)) {
      continue;
    }
    for (const fact of unitFacts) {
      rows.push({
        concept,
        value: fact.val ?? null,
        unit,
        periodEnd: fact.start ? fact.end || null : null,
        periodInstant: fact.start ? null : fact.end || null,
        end: fact.end ?? null,
        filed: fact.filed ?? null,
        form: fact.form ?? null,
      });
    }
  }

  rows.sort((a, b) => {
    const endCompare = (a.end || "").localeCompare(b.end || "");
    if (endCompare !== 0) {
      return endCompare;
    }
    return (a.filed || "").localeCompare(b.filed || "");
  });
  return rows;
}

File: python/src/sec_edgar_toolkit/core/xbrl/queries.py
"""
Fact-level query primitives over the company-facts payload.

- ``FactsData`` — the raw company-facts payload with a history helper
- ``FactQuery`` — a list of fact records with chainable helpers
- ``build_fact_records`` — normalize one concept's raw facts into records
- ``parse_filter_expression`` — parse "concept=Assets&unit=USD" strings
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Tuple


class FactsData(dict):
    """Raw company-facts payload with a concept-history helper."""

    def facts_history(self, concept: str, taxonomy: str = "us-gaap"):
        """
        History of one concept as a pandas DataFrame with columns
        ``value, unit, period_end, period_instant, filed, form``.
        """
        import pandas as pd

        concept_data = (self.get("facts", {}) or {}).get(taxonomy, {}).get(concept)
        if not concept_data:
            return pd.DataFrame()

        rows = []
        for unit, unit_facts in (concept_data.get("units") or {}).items():
            for fact in unit_facts:
                start = fact.get("start")
                end = fact.get("end")
                rows.append(
                    {
                        "concept": concept,
                        "value": fact.get("val"),
                        "unit": unit,
                        "period_end": end if start else None,
                        "period_instant": None if start else end,
                        "end": end,
                        "filed": fact.get("filed"),
                        "form": fact.get("form"),
                    }
                )
        df = pd.DataFrame(rows)
        if not df.empty:
            df = df.sort_values(by=["end", "filed"], na_position="first").reset_index(
                drop=True
            )
        return df


class FactQuery(list):
    """Query result: a list of fact records with chainable helpers."""

    def by_concept(self, concept: str) -> "FactQuery":
        """Narrow the results to a single concept name."""
        needle = concept.lower()
        return FactQuery(
            record
            for record in self
            if needle in str(record.get("concept", "")).lower()
        )

    def to_dataframe(self):
        """Results as a pandas DataFrame."""
        import pandas as pd

        return pd.DataFrame(list(self))


def parse_filter_expression(
    expression: str, taxonomy: str, unit: Optional[str], period: Optional[str]
) -> Tuple[Optional[str], str, Optional[str], Optional[str]]:
    """
    Parse a "key=value&key=value" filter expression.

    Returns the (concept, taxonomy, unit, period) tuple with explicit
    keyword arguments taking precedence over the expression.
    """
    parsed: Dict[str, str] = {}
    for part in expression.split("&"):
        if "=" in part:
            key, _, value = part.partition("=")
            parsed[key.strip()] = value.strip()
    return (
        parsed.get("concept"),
        parsed.get("taxonomy", taxonomy),
        unit or parsed.get("unit"),
        period or parsed.get("period"),
    )


def build_fact_records(
    concept_name: str,
    concept_data: Dict[str, Any],
    taxonomy: str,
    unit_filter: Optional[str] = None,
    period_filter: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """Normalize one concept's raw company-facts entries into fact records."""
    results: List[Dict[str, Any]] = []

    units = concept_data.get("units", {})
    for unit, unit_data in units.items():
        if unit_filter and unit != unit_filter:
            continue

        for fact in unit_data:
            if period_filter:
                fact_period = fact.get("fy") or fact.get("fp") or fact.get("frame", "")
                if period_filter not in str(fact_period):
                    continue

            start = fact.get("start")
            end = fact.get("end")
            results.append(
                {
                    "concept": concept_name,
                    "taxonomy": taxonomy,
                    "value": fact.get("val"),
                    "unit": unit,
                    "period": fact.get("frame")
                    or f"FY{fact.get('fy', '')}{fact.get('fp', '')}",
                    "fiscal_year": fact.get("fy"),
                    "fiscal_period": fact.get("fp"),
                    "start_date": start,
                    "end_date": end,
                    "period_end": end if start else None,
                    "period_instant": None if start else end,
                    "context": fact.get("accn"),
                    "filed": fact.get("filed"),
                    "accession_number": fact.get("accn"),
                    "form": fact.get("form"),
                }
            )

    return results
